chore(app,app-shell,app-shell-odd, protocol-designer, labware-library, components, labware-designer): update Electron version to v27.0.0

koji opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
Update Electron version 27.0.0
https://www.electronjs.org/docs/latest/tutorial/electron-timelines#timeline

There is `loadURL` issue in the versions that include the following PR.
https://github.com/electron/electron/pull/40143

At this moment we cannot update any newer version. (I'll keep checking the repo)


This PR also includes Nodejs update because of Electron v27's requirement.
This PR needs Nodejs v21.4.0 since M1 Mac has an issue about running Desktop app. (I'll keep an eye on this to support more nodejs version). M2 Mac can use nodejs v18.

Need `NODE_OPTION` due to OpenSSL version.
I tried to export NODE_OPTION in the top of Makefile, but somehow that has an impact on other part and prevent building the app(Desktop/ODD). I use `NODE_OPTION` separately.
Another reasons for this are `labware-library` and `protocol-designer` since the option prevents Cypress.


### Running the application
node v21
- [x] macOS
- [x] Windows
- [x] Ubuntu
- [x] ODD mode
- [x] ODD

### CI
- [x] update Nodejs version (v16 -> v21)

### OT-2
- [x] USB connection

### Other applications
- [x] Protocol Designer
- [x] Labware Library
- [x] Labware Designer
- [x] Storybook

### Commands
- [x] make check-js
- [x] make lint-js
- [x] make test-js

This PR removes `usb-detection` since the repo has been archived and not compatible with node v18+.
The new package we need to use is `[node-usb](https://github.com/node-usb/node-usb)`
https://working-wormhole-d95.notion.site/usb-detection-node-usb-a6637f72f5d44f3ba9c7520b4f0da24f

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Test Plan
- Run Desktop app and check that the app works as well as the current edge
- Run ODD app and check that the app works as well as the current edge
<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

koji committed:
chore(app,app-shell,app-shell-odd): update Electron version to v23

Update the Electron version from v21 to v23. Actually v23 is over the life but ODD has compatibility
issues with Chromium v112 and later

RAUT-

koji committed:
update Nodejs version in CI files

koji committed:
add node_option to labware designer

koji committed:
update ui deprecated part and remove usb-detection

koji committed:
updatet app-shell Makefile

koji committed:
Windows works

koji committed:
use v27.0.0 instead of the end of life versions

koji committed:
add node option to Makefile in components

koji committed:
Merge branch 'edge' into chore_update-electron-v27.0.0

koji committed:
fix odd mode and update electron-context-menu

koji committed:
remove webcontent.on new-window

koji committed:
for building asar for ODD

koji committed:
update node version to 20

koji committed:
Update Makefile

koji committed:
remove usb-detection related codes

koji committed:
update type for tests

koji committed:
update tests

koji committed:
fix type issues

koji committed:
fix test case to use usb package

koji committed:
change ci's nodejs version to v21

koji committed:
Merge branch 'edge' into chore_update-electron-v27.0.0

koji committed:
Update dispatch.test.ts

koji committed:
Update dispatch.test.ts

koji committed:
remove export

koji committed:
fix type issues

koji committed:
fix secondary button test

koji commented:
- [ ] check redux

koji committed:
chnage the way to access getDevices method

koji committed:
fix test-js and check-js errors

koji committed:
update Makefile to fix ci error on cypress

koji commented:
- [ ] fix cypress issue 
  - [ ] labware-library
  - [ ] protocol-designer

koji committed:
Update Makefile

koji committed:
pass option separately due to cypress

brenthagen commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -56,7 +56,7 @@ jobs:
       - uses: 'actions/setup-node@v3'
         with:
-          node-version: '16'
+          node-version: '21'
```
do we need to use 21, why not 20? Node uses an odd/even release pattern. even releases have long-term support and are recommended for production: https://nodejs.org/en/about/previous-releases

koji replied:
Ed tried to v20 and it didn't work on his end. But v21 worked on his end.

mjhuff replied:
Did we ever characterize Ed's error? I think trying to use 20 over 21 for the sake of LTS would be better if possible.

koji replied:
There was no specific error which was the same as node v18 + M1 macOS.
logs about `shell` don't show up.

brenthagen commented on package.json:
```diff
@@ -27,7 +27,7 @@
   "packageManager": "yarn@1.22.19",
   "engines": {
-    "node": "^16.9.0"
+    "node": ">=18.14.0 <22.0.0"
```
should we pin to a version? I'm using `20.11.0`

koji replied:
yeah, fixing version would be safe.

sfoster1 replied:
Can we actually not do that? For the following reasons:
- This version of node is _not_ the version of node that will be used inside electron; electron has its own node binary that runs at a specific version, which for 27 is 18.19 (see https://www.electronjs.org/docs/latest/tutorial/electron-timelines) and presumably patched to work on those osx versions
- the node that we specify here is therefore actually only run for executing dev tooling
   - we should probably run tests under the same version as electron, and should therefore prefer 18.19
   - but sometimes we can't so we allow the range

brenthagen reviewed (commented):
I'm getting some strange behavior on this branch from locally simulated robots - reporting as unreachable but responding with 200s to various requests. I'm not sure why, digging a bit further.

<img width="1136" alt="Screen Shot 2024-01-26 at 11 25 47 AM" src="https://github.com/Opentrons/opentrons/assets/29845468/11bb91ce-33e9-4fe4-9585-2937e35e8fcf">

<img width="942" alt="Screen Shot 2024-01-26 at 11 28 27 AM" src="https://github.com/Opentrons/opentrons/assets/29845468/5b9270ce-a78b-4ac1-9966-2c575302d1bd">

sfoster1 commented:
Looks like flex builds will fail until we get node-native on >=18 https://github.com/Opentrons/oe-core/actions/runs/7659498340/job/20874747633

sfoster1 commented:
Tracking the oe-core work here https://github.com/Opentrons/oe-core/pull/128

sfoster1 commented:
That build succeeded: https://builds.opentrons.com/ot3-oe/7672229872/ot3-system.zip

sfoster1 commented:
And it works, happy with this from the ODD side

koji commented:
- [ ] try v18.19.0 again
 - [ ] solve conflicts

koji committed:
address feedback use nodejs v18.19.0

koji committed:
Merge branch 'edge' into chore_update-electron-v27.0.0

koji committed:
fix check-js errors

koji committed:
fix test-js error

koji committed:
update Makefile for CI tests

koji committed:
update for CI test

koji committed:
Merge branch 'edge' into chore_update-electron-v27.0.0

koji committed:
Update yarn.lock

koji committed:
change node version to 18.19.0

brenthagen committed:
set ipv4first dns default result order

brenthagen commented:
> I'm getting some strange behavior on this branch from locally simulated robots - reporting as unreachable but responding with 200s to various requests. I'm not sure why, digging a bit further.

this is an issue with node-fetch and dns default result order https://opentrons.slack.com/archives/CSCLVUW3C/p1706568971026189

i've pushed a commit to fix

brenthagen commented on app-shell-odd/src/main.ts:
```diff
@@ -27,6 +28,15 @@ import { registerNotify, closeAllNotifyConnections } from './notify'
+ * setting the default to IPv4 fixes the issue
+ * https://github.com/node-fetch/node-fetch/issues/1624
+ */
+// TODO(bh, 2024-1-30): @types/node needs to be updated to address this type error. updating @types/node will also require updating our typescript version
```
this is failing because our wildcard @types/node version is 15 in yarn.lock. i'm looking into updating our wildcard `@types/node` version and typescript in a separate commit.

brenthagen committed:
Merge branch 'edge' into chore_update-electron-v27.0.0

brenthagen committed:
fix yaml lint

sfoster1 committed:
fix: dev servers must serve on all protocols

2a360ce6a6efa3f2e423a3404193ae4d0738212a tells node to prefer ipv4
addresses since in node >17 it prefers ipv6. Unfortunately, when we're
running cypress we don't really have the ability to specify this since
it happens in code. Instead, we can make sure that webpack-dev-server
serves on ipv6 as well as ipv4; the right way to do this is to tell it
to use the host :: which is ipv6 shorthand for "accept-all-incoming" in
the same way that 0.0.0.0 is for ipv4.

This should actually accept all incoming v4 requests too. You can test
it by running make dev and then doing both
curl --head 127.0.0.1:8080/ (the ipv4 version)
curl --head '[::1]:8080/' (the ipv6 version; ::1 is ipv6 localhost, and
the brackets are how ipv6 is written. if doing this in the shell you'll
need single-quotes here)
curl --head http://localhost:8080/

sfoster1 commented on labware-library/Makefile:
```diff
@@ -22,17 +22,18 @@ clean:
 dev: export NODE_ENV := development
 dev:
-	webpack-dev-server --hot
+	export NODE_OPTIONS=--openssl-legacy-provider && webpack-dev-server --hot --host=::
```
the other side of the ipv4/v6 thing is that it also applies to cypress runs, and there we can't run our own js code, so let's just serve on v6 wildcard. you can test this by doing `curl --head '[::1]:8080/'` (note the single quotes - your shell will try and parse those brackets) when dev server is running, as well as `curl --head 127.0.0.1:8080/` and `curl --head http://localhost:8080/`

brenthagen replied:
think we need this in `app` also, i get same behavior locally on node 18 (but not 20 it seems).

and i guess labware-designer while we're at it

sfoster1 committed:
get some timing info on cypress

sfoster1 commented:
- [x] fix ci errors

sfoster1 committed:
add v6 bindings to most all dev servers

and twiddle some env exports a bit

sfoster1 reviewed (approved):
Looks good to me! @koji I think you should have the honor of clicking merge!
